File: tasks/16_legacy_java_expense_report_refactor/expected_solution/src/expenseReport.ts
export type EmployeeLevel = "STAFF" | "MANAGER" | "EXEC" | "CONTRACTOR";
export type ExpenseType = "MEAL" | "LODGING" | "TRAVEL" | "SUPPLIES";

export interface Employee {
  level: EmployeeLevel;
  monthlyLimitCents: number;
}

export interface LineItem {
  type: ExpenseType;
  amountCents: number;
  city: string;
  hasReceipt: boolean;
}

export interface Decision {
  status: "APPROVED" | "REVIEW" | "DENIED";
  approver: "NONE" | "MANAGER" | "FINANCE";
  reasons: string[];
}

export function decide(employee: Employee, items: LineItem[]): Decision {
  let total = 0;
  const reasons: string[] = [];

  for (const item of items) {
    total += item.amountCents;
    const mealLimit = item.city === "NYC" || item.city === "SFO" ? 7500 : 5000;
    if (item.type === "MEAL" && item.amountCents > mealLimit) reasons.push("meal-limit");
    if (!item.hasReceipt && item.amountCents > 2500 && employee.level !== "EXEC") {
      reasons.push("missing-receipt");
    }
    if (item.type === "LODGING" && item.amountCents > 40000) reasons.push("lodging-limit");
  }

  if (total > employee.monthlyLimitCents) reasons.push("monthly-limit");
  if (reasons.length > 0) return { status: "DENIED", approver: "MANAGER", reasons };
  if (total >= 100000 || employee.level === "CONTRACTOR") {
    return { status: "REVIEW", approver: "FINANCE", reasons };
  }
  return { status: "APPROVED", approver: "NONE", reasons };
}

File: tasks/16_legacy_java_expense_report_refactor/expected_solution/legacy/ExpenseReport.java
import java.util.*;

class ExpenseReport {
  Decision decide(Employee employee, List<LineItem> items) {
    int total = 0;
    List<String> reasons = new ArrayList<>();
    for (LineItem item : items) {
      total += item.amountCents;
      int mealLimit = item.city.equals("NYC") || item.city.equals("SFO") ? 7500 : 5000;
      if (item.type.equals("MEAL") && item.amountCents > mealLimit) reasons.add("meal-limit");
      if (!item.hasReceipt && item.amountCents > 2500 && !employee.level.equals("EXEC")) reasons.add("missing-receipt");
      if (item.type.equals("LODGING") && item.amountCents > 40000) reasons.add("lodging-limit");
    }
    if (total > employee.monthlyLimitCents) reasons.add("monthly-limit");
    if (!reasons.isEmpty()) return new Decision("DENIED", "MANAGER", reasons);
    if (total >= 100000 || employee.level.equals("CONTRACTOR")) return new Decision("REVIEW", "FINANCE", reasons);
    return new Decision("APPROVED", "NONE", reasons);
  }
}
